pipeline/lib/api.ml

```diff
@@ -1,6 +1,12 @@
 open Lwt.Infix
 module Utils = Cohttp_lwt_unix
 
+exception AuthError of string
+
+exception PermissionError of string
+
+exception ServerError of string
```

Comment:
This is a bit subjective and you'll have to choose what you prefer, but it can be helpful to define more domain specific exceptions. Here you are in a middle ground between the HTTP errors and the domain... and so you need `of string` (which to the future reader means "I will reuse this exception for different cause of failures")

Defining exceptions acts as a warning "here is what can go wrong":
```ocaml
exception Api_not_enabled
exception Invalid_token
exception Wrong_token
```

This avoids the redundant `AuthError "Invalid token"` and mean all the user facing error messages can be defined in one place, independently of the logic.

But in this specific situation, you could also go all in with the HTTP errors in order to simplify the unique `try...with`:

```ocaml
exception Error of [`Bad_request | `Unauthorized | ... ] * string
let invalid_token () = raise (Error (`Unauthorized, "Invalid token"))
let wrong_token () = raise (Error ...)
```

(I like this less as it makes it harder to add unit tests on the validation, but it's less verbose and achieves the same result)

Reply:
Thanks for the clear explanation. Domain specific exceptions make sense, thanks!